### ISSUE TITLE
Navigation to c reference with outdated compile-commands.json causes out-of-bounds panic. 

### ISSUE THREAD

**Issue by AloeareV** [helix-editor/helix#934]
### Reproduction steps
After building compile-commands.json, `mv foo.h foo.hpp`, then `hx foo.hpp` and use `gr` to navigate to reference in same file. Helix tries to read `foo.h`, which no longer exists, and panics with 
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Line index out of bounds: line index 319, Rope/RopeSlice line count 2', /home/hazelgrace/.cargo/registry/src/github.com-1ecc6299db9ec823/ropey-1.3.1/src/slice.rs:338:41
```

### Environment

- Platform: Arch Linux

- Helix version: v0.5.0-13-ge5de103 
 
<details><summary>~/.cache/helix/helix.log</summary>

```
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] clangd version 12.0.1\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] PID: 2166428\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] Working directory: /home/hazelgrace/zcashes/zip155-active\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] argv[0]: clangd\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] Starting LSP over stdin/stdout\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] <-- initialize(0)\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] --> reply:initialize(0) 0 ms\n"
2021-10-29T14:37:13.924 helix_lsp::transport [ERROR] err <- "I[14:37:13.924] <-- initialized\n"
2021-10-29T14:37:13.925 helix_lsp::transport [ERROR] err <- "I[14:37:13.925] <-- textDocument/didOpen\n"
2021-10-29T14:37:13.929 helix_lsp::transport [ERROR] err <- "I[14:37:13.929] Loaded compilation database from /home/hazelgrace/zcashes/zip155-active/compile_commands.json\n"
2021-10-29T14:37:13.929 helix_lsp::transport [ERROR] err <- "I[14:37:13.929] ASTWorker building file /home/hazelgrace/zcashes/zip155-active/src/serialize.hpp version 0 with command inferred from /home/hazelgrace/zcashes/zip155-active/src/test/serialize_tests.cpp\n"
2021-10-29T14:37:13.929 helix_lsp::transport [ERROR] err <- "[/home/hazelgrace/zcashes/zip155-active/src]\n"
2021-10-29T14:37:13.929 helix_lsp::transport [ERROR] err <- "/home/hazelgrace/zcashes/zip155-active/depends/x86_64-pc-linux-gnu/native/bin/clang++ --driver-mode=g++ -target x86_64-pc-linux-gnu -B/home/hazelgrace/zcashes/zip155-active/depends/x86_64-pc-linux-gnu/native/bin -stdlib=libc++ -m64 -DHAVE_CONFIG_H -I. -I../src/config -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I. -I./obj -pthread -I/home/hazelgrace/zcashes/zip155-active/depends/x86_64-pc-linux-gnu/share/../include -I./leveldb/include -I./leveldb/helpers/memenv -I./rust/include -I./secp256k1/include -I./univalue/include -I./test/ -I/home/hazelgrace/zcashes/zip155-7257-7868-7749_ee06e0/depends/x86_64-pc-linux-gnu/include -Qunused-arguments -I/home/hazelgrace/zcashes/zip155-active/depends/x86_64-pc-linux-gnu/share/../include/ -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -g -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -pipe -O3 -fwrapv -fno-strict-aliasing -c -std=c++17 /home/hazelgrace/zcashes/zip155-active/src/serialize.hpp -fsyntax-only -resource-dir=/usr/lib/clang/12.0.1\n"
2021-10-29T14:37:13.930 helix_lsp::transport [ERROR] err <- "I[14:37:13.930] --> window/workDoneProgress/create(0)\n"
2021-10-29T14:37:13.930 helix_lsp::transport [ERROR] err <- "I[14:37:13.930] Enqueueing 292 commands for indexing\n"
2021-10-29T14:37:13.930 helix_lsp::transport [ERROR] err <- "I[14:37:13.930] <-- reply(0)\n"
2021-10-29T14:37:13.930 helix_lsp::transport [ERROR] err <- "I[14:37:13.930] --> $/progress\n"
2021-10-29T14:37:13.930 helix_lsp::transport [ERROR] err <- "I[14:37:13.930] --> $/progress\n"
2021-10-29T14:37:14.432 helix_lsp::transport [ERROR] err <- "E[14:37:14.432] tidy-config error at /home/hazelgrace/zcashes/zip155-active/.clang-tidy:1:0: not a mapping\n"
2021-10-29T14:37:14.432 helix_lsp::transport [ERROR] err <- "E[14:37:14.432] Error parsing clang-tidy configuration in /home/hazelgrace/zcashes/zip155-active/.clang-tidy: Invalid argument\n"
2021-10-29T14:37:14.475 helix_lsp::transport [ERROR] err <- "I[14:37:14.475] --> textDocument/publishDiagnostics\n"
2021-10-29T14:37:14.735 helix_lsp::transport [ERROR] err <- "E[14:37:14.735] Background-index: Couldn't read /home/hazelgrace/zcashes/zip155-active/src/serialize.h to validate stored index: No such file or directory\n"
2021-10-29T14:37:14.752 helix_lsp::transport [ERROR] err <- "I[14:37:14.752] --> $/progress\n"
2021-10-29T14:37:14.752 helix_lsp::transport [ERROR] err <- "I[14:37:14.752] --> $/progress\n"
2021-10-29T14:37:19.857 helix_lsp::transport [ERROR] err <- "I[14:37:19.857] <-- textDocument/references(1)\n"
2021-10-29T14:37:19.861 helix_lsp::transport [ERROR] err <- "I[14:37:19.861] --> reply:textDocument/references(1) 4 ms\n"
```

</details>

**Comment by AloeareV** [comment]
I stumbled across this trying to get clangd to interpret `foo.h` as a `c++` file, as it was interpreting it as a `c` file and throwing errors like `undefined typename 'template'` for a templated function.

**Comment by Houkime** [comment]
*NOTE:* this error is NOT triggered by just having a nonexistent file in the compile_commands and using `gr`.

* unpack this [foo.zip](https://github.com/helix-editor/helix/files/7445308/foo.zip)
* `bear -- gcc -I. -o main main.c other.c` to generate compile_commands ([bear](https://github.com/rizsotto/Bear))
* `hx main.c`
* `gr` on bar(). See it works.
* exit helix
* `rm other.c` (a file with bar() definition)
* `hx main.c`
* `gr` on bar(). See this:

![lspbug](https://user-images.githubusercontent.com/32805122/139509108-7121dc4d-04bb-40c1-b60d-5fab6df00a1d.png)

LSP still returns other.c as existing, but no crash, and if it is selected and edited, it is just an empty file (that will be created on save)

**Comment by Houkime** [comment]
if you then roll back (using git for example, or by reunzipping), and, without remaking compilation-commands do renaming:
* `rename \.c .cpp ./*.c`
* `rename \.h .hpp ./*.h`
and then 
* `hx main.c`
* `gr` on bar()
Nothing will happen, bar() will be just selected as the sole reference.

**Comment by Houkime** [comment]
same for cpp if other.cpp was removed (so it is not langua-ge dependent)
![screen](https://user-images.githubusercontent.com/32805122/139510556-a14264b0-d086-42b4-927e-508ac1c1288d.png)

**Comment by Houkime** [comment]
If one makes a slight adjustment, and makes it so that bar() is NOT on the line 0 of other.cpp, like this:
![screen](https://user-images.githubusercontent.com/32805122/139510797-46fb0dc3-5191-486b-81e5-c6451ee7b774.png)

even then it won't crash.
![screen](https://user-images.githubusercontent.com/32805122/139510956-771069a3-4cf9-4405-9e70-342e8ced98f8.png)

**Comment by Houkime** [comment]
But, if we make the line number high enough (more than fits on screen)
![screen](https://user-images.githubusercontent.com/32805122/139511283-c038a0d2-5052-47ed-8db0-f9023e200577.png)

gr works as usual...

![screen](https://user-images.githubusercontent.com/32805122/139511494-1f95a07b-d705-42eb-a2c4-62a93d5ab361.png)

Until we navigate to the line of the nonexistent file. THEN it will crash.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Line index out of bounds: line index 151, Rope/RopeSlice line count 2', /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/ropey-1.3.1/src/slice.rs:338:41
stack backtrace:
   0: rust_begin_unwind
             at /build/rust/src/rustc-1.56.0-src/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /build/rust/src/rustc-1.56.0-src/library/core/src/panicking.rs:101:14
   2: core::result::unwrap_failed
             at /build/rust/src/rustc-1.56.0-src/library/core/src/result.rs:1617:5
   3: ropey::slice::RopeSlice::line_to_byte
   4: helix_term::ui::editor::EditorView::doc_syntax_highlights
   5: <helix_term::ui::picker::FilePicker<T> as helix_term::compositor::Component>::render
   6: helix_term::compositor::Compositor::render
   7: helix_term::application::Application::handle_terminal_events
   8: helix_term::application::Application::run::{{closure}}
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  10: tokio::park::thread::CachedParkThread::block_on
  11: tokio::runtime::thread_pool::ThreadPool::block_on
  12: tokio::runtime::Runtime::block_on
  13: hx::main
```

**Comment by Houkime** [comment]
Ok, so here is the **actual reproduction project**.

* unpack this [foo.zip](https://github.com/helix-editor/helix/files/7445396/foo.zip)
* `bear -- gcc -I. -o main main.cpp other.cpp` to generate compile_commands ([bear](https://github.com/rizsotto/Bear))
* `hx main.cpp`
* `gr` on bar(). See it works.
* exit helix
* `rm other.cpp` (a file with bar() definition)
* `RUST_BACKTRACE=1 hx main.cpp`
* `gr` on bar(). 
* navigate to a nonexisting file other.cpp .
* See error (stack in the post above above)

**Comment by Houkime** [comment]
Proposed solution: make pickers more robust.
Variants (combinable):
 * check for file existence before rendering preview.
 * check for line count vs lsp suggested line. Ignore the entry if out of bounds.
Ideally - both.

**Comment by AloeareV** [comment]
Good lord, thanks for tracking this down! I just assumed it would generalize more than it did, if I'd attempted to minimally replicate it I don't know that I ever would have thought of being far down in the file being a prereq.

**Comment by AloeareV** [comment]
I believe this was fixed by #1207
